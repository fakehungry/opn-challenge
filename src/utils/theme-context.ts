import { Dispatch, SetStateAction, createContext } from 'react';
import lightTheme from '../styles/light-theme';

export enum ThemeMode {
  LIGHT = 'light',
  DARK = 'dark',
}

type ThemeContextType = {
  themeMode: ThemeMode;
  theme: typeof lightTheme;
  setThemeMode: Dispatch<SetStateAction<ThemeMode | null>>;
};

export const ThemeContext = createContext<ThemeContextType>({
  themeMode: ThemeMode.LIGHT,
  theme: lightTheme,
  setThemeMode: () => {},
});
