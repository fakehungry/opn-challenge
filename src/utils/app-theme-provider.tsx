import { useEffect, useState } from 'react';
import { ThemeContext, ThemeMode } from './theme-context';
import lightTheme from '../styles/light-theme';
import { ThemeProvider } from 'styled-components';
import { theme } from '../styles/styled-component-theme';
import darkTheme from '../styles/dark-theme';

type AppThemeProviderProps = {
  children: JSX.Element | JSX.Element[];
  theme?: ThemeMode;
};

const getThemeFromLocalStorage = () =>
  window.localStorage.getItem('theme') || ThemeMode.LIGHT;

const setThemeToLocalStorage = (theme: ThemeMode) =>
  window.localStorage.setItem('theme', theme);

export const AppThemeProvider = (props: AppThemeProviderProps) => {
  const { children, theme: themeProp } = props;
  const [themeMode, setThemeMode] = useState<ThemeMode | null>(null);
  const colors = themeMode === ThemeMode.LIGHT ? lightTheme : darkTheme;

  useEffect(() => {
    if (themeProp) {
      setThemeMode(themeProp);
    } else if (themeMode === null) {
      setThemeMode(getThemeFromLocalStorage() as ThemeMode);
      return;
    }

    setThemeToLocalStorage(themeMode as ThemeMode);
  }, [themeMode, themeProp]);

  return (
    <ThemeContext.Provider
      value={{ themeMode: themeMode as ThemeMode, theme: colors, setThemeMode }}
    >
      <ThemeProvider theme={{ ...theme, colors: colors }}>
        {children}
      </ThemeProvider>
    </ThemeContext.Provider>
  );
};
