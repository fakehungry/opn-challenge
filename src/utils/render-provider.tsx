import React from 'react';
import { render } from '@testing-library/react';
import { AppThemeProvider } from './app-theme-provider';
import { ThemeMode } from './theme-context';

// REMARK: Test only light mode
export const renderProvider = (props: any) => {
  return render(
    <AppThemeProvider theme={ThemeMode.LIGHT}>{props}</AppThemeProvider>,
  );
};
