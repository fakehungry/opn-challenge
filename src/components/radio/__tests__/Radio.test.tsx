import { fireEvent } from '@testing-library/react';
import Radio from '..';
import { renderProvider } from '../../../utils/render-provider';

describe('radio component', () => {
  it('should render radio with label and match previous snapshot', () => {
    const { container } = renderProvider(<Radio label="Radio" />);

    expect(container).toMatchSnapshot();
  });

  it('should change background color on checked', () => {
    const { getByRole } = renderProvider(<Radio label="Radio" />);

    fireEvent.click(getByRole('radio'));
    expect(getByRole('radio')).toBeChecked();
    expect(getByRole('radio')).toHaveStyle('background-color: #0070f3');
  });
});
