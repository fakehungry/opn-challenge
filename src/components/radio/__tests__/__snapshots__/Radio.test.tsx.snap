// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`radio component should render radio with label and match previous snapshot 1`] = `
<div>
  <label
    class="sc-guDLey cmzxpj"
  >
    <input
      class="sc-beySPh kjPIbG"
      type="radio"
    />
    <span>
      Radio
    </span>
  </label>
</div>
`;
